File: packages/common/src/finance_common/parsing/statement_import_chart_kinds.py
"""Chart kind resolution for statement import (mirrors dashboard logic)."""

from __future__ import annotations

REFUND_MATCH_TOLERANCE_INR = 15


def _tx_kind(row: dict) -> str:
    return str(row.get("tx_kind") or "spend").lower()


def _amounts_match(spend_amount: float, refund_amount: float) -> bool:
    diff = abs(abs(spend_amount) - abs(refund_amount))
    return diff <= REFUND_MATCH_TOLERANCE_INR


def _match_score(spend: dict, refund: dict) -> float:
    diff = abs(abs(float(spend["amount"])) - abs(float(refund["amount"])))
    same_period = 0 if spend.get("statement_period") == refund.get("statement_period") else 1
    date_ok = 0 if spend["date"] <= refund["date"] else 1
    return same_period * 1000 + date_ok * 100 + diff


def resolve_chart_kind_rows(rows: list[dict]) -> list[dict]:
    spends = [r for r in rows if _tx_kind(r) == "spend"]
    refunds = [r for r in rows if _tx_kind(r) == "refund"]
    others = [r for r in rows if _tx_kind(r) not in ("spend", "refund")]

    used_spend_ids: set[str] = set()
    offset_refund_ids: set[str] = set()

    for refund in refunds:
        best: tuple[dict, float] | None = None
        for spend in spends:
            if spend["id"] in used_spend_ids:
                continue
            if spend["bank"] != refund["bank"] or spend["card"] != refund["card"]:
                continue
            if not _amounts_match(float(spend["amount"]), float(refund["amount"])):
                continue
            score = _match_score(spend, refund)
            if best is None or score < best[1]:
                best = (spend, score)
        if best:
            used_spend_ids.add(best[0]["id"])
            offset_refund_ids.add(refund["id"])

    result: list[dict] = []
    for t in spends:
        if t["id"] in used_spend_ids:
            continue
        result.append({"id": t["id"], "kind": "spend", "amount": abs(float(t["amount"]))})
    for t in refunds:
        if t["id"] in offset_refund_ids:
            continue
        result.append({"id": t["id"], "kind": "payment", "amount": abs(float(t["amount"]))})
    for t in others:
        result.append({"id": t["id"], "kind": _tx_kind(t), "amount": abs(float(t["amount"]))})
    return result


def aggregate_chart_kinds_by_type(rows: list[dict]) -> dict[str, dict[str, float | int]]:
    sums: dict[str, dict[str, float | int]] = {}
    for row in resolve_chart_kind_rows(rows):
        kind = row["kind"]
        if kind not in sums:
            sums[kind] = {"count": 0, "amount": 0.0}
        sums[kind]["count"] = int(sums[kind]["count"]) + 1
        sums[kind]["amount"] = float(sums[kind]["amount"]) + float(row["amount"])
    return sums

File: dashboard/src/lib/statementImportChartKinds.ts
import type { StatementImportTransactionRow } from '@/types/api'

/** Max ₹ difference when pairing a refund to an expense (discount rounding). */
export const REFUND_MATCH_TOLERANCE_INR = 15

export type ChartKindRow = {
  id: string
  kind: string
  amount: number
}

function txKind(t: StatementImportTransactionRow): string {
  return (t.tx_kind ?? 'spend').toLowerCase()
}

function amountsMatch(spendAmount: number, refundAmount: number): boolean {
  const diff = Math.abs(Math.abs(spendAmount) - Math.abs(refundAmount))
  return diff <= REFUND_MATCH_TOLERANCE_INR
}

function matchScore(
  spend: StatementImportTransactionRow,
  refund: StatementImportTransactionRow,
): number {
  const diff = Math.abs(Math.abs(spend.amount) - Math.abs(refund.amount))
  const samePeriod = spend.statement_period === refund.statement_period ? 0 : 1
  const dateOk = spend.date <= refund.date ? 0 : 1
  return samePeriod * 1000 + dateOk * 100 + diff
}

/**
 * Prepare rows for type charts:
 * - Matched refund + expense pairs are excluded (netted off).
 * - Unmatched refunds are treated as bill paid.
 * - Refund is never a chart category.
 */
export function resolveChartKindRows(
  rows: StatementImportTransactionRow[],
): ChartKindRow[] {
  const spends = rows.filter((t) => txKind(t) === 'spend')
  const refunds = rows.filter((t) => txKind(t) === 'refund')
  const others = rows.filter((t) => {
    const k = txKind(t)
    return k !== 'spend' && k !== 'refund'
  })

  const usedSpendIds = new Set<string>()
  const offsetRefundIds = new Set<string>()

  for (const refund of refunds) {
    let best: { spend: StatementImportTransactionRow; score: number } | null = null
    for (const spend of spends) {
      if (usedSpendIds.has(spend.id)) continue
      if (spend.bank !== refund.bank || spend.card !== refund.card) continue
      if (!amountsMatch(spend.amount, refund.amount)) continue
      const score = matchScore(spend, refund)
      if (!best || score < best.score) {
        best = { spend, score }
      }
    }
    if (best) {
      usedSpendIds.add(best.spend.id)
      offsetRefundIds.add(refund.id)
    }
  }

  const result: ChartKindRow[] = []

  for (const t of spends) {
    if (usedSpendIds.has(t.id)) continue
    result.push({ id: t.id, kind: 'spend', amount: Math.abs(t.amount) })
  }

  for (const t of refunds) {
    if (offsetRefundIds.has(t.id)) continue
    result.push({ id: t.id, kind: 'payment', amount: Math.abs(t.amount) })
  }

  for (const t of others) {
    result.push({ id: t.id, kind: txKind(t), amount: Math.abs(t.amount) })
  }

  return result
}

export function aggregateChartKindsByType(
  rows: StatementImportTransactionRow[],
): Record<string, { count: number; amount: number }> {
  const sums: Record<string, { count: number; amount: number }> = {}
  for (const row of resolveChartKindRows(rows)) {
    if (!sums[row.kind]) sums[row.kind] = { count: 0, amount: 0 }
    sums[row.kind].count += 1
    sums[row.kind].amount += row.amount
  }
  return sums
}
